controllers/tokenupload_controller.go

```diff
@@ -0,0 +1,235 @@
+					continue
+				} else {
+					// this is the only place where we can get the name of just created SPIAccessToken
+					// which is presumably OK since SPI (binding) controller will look for the token by type/URL ?
+					lg.V(logs.DebugLevel).Info("SPI Access Token created : " + accessToken.Name)
+				}
+			} else {
+				accessToken = *tkn
+				lg.V(logs.DebugLevel).Info("SPI Access Token found by providerUrl : " + accessToken.Name + " nothing changed.")
+			}
+
+		} else {
+			logError(ctx, s, fmt.Errorf("Secret is invalid, it is labeled with %s but neither %s nor %s label provided: %w", tokenSecretLabel, spiTokenNameLabel, providerUrlLabel, err), r, lg)
+			continue
+		}
+
+		token := spi.Token{
+			Username:    string(s.Data["userName"]),
+			AccessToken: string(s.Data["tokenData"]),
+		}
+
+		logs.AuditLog(ctx).Info("manual token upload initiated", s.Namespace, accessToken.Name)
+		err = r.TokenStorage.Store(ctx, &accessToken, &token)
+		if err != nil {
+			logError(ctx, s, fmt.Errorf("token storing failed: %w", err), r, lg)
+			logs.AuditLog(ctx).Error(err, "manual token upload failed", s.Namespace, accessToken.Name)
+			continue
+		}
+		logs.AuditLog(ctx).Info("manual token upload completed", s.Namespace, accessToken.Name)
+
+		tryDeleteEvent(ctx, s.Name, req.Namespace, r, lg)
+
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *TokenUploadReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Secret{}).
+		WithEventFilter(createTokenPredicate()).
+		Complete(r)
+}
+
+func createTokenPredicate() predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			_, ok := e.Object.GetLabels()[tokenSecretLabel]
+			return ok
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			_, newLabelExists := e.ObjectNew.GetLabels()[tokenSecretLabel]
+			_, oldLabelExists := e.ObjectOld.GetLabels()[tokenSecretLabel]
+			if newLabelExists && !oldLabelExists {
+				return true
+			}
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return false
+		},
+	}
+}
+
+// reports error in both Log and Event in current Namespace
+func logError(ctx context.Context, secret corev1.Secret, err error, r *TokenUploadReconciler, lg logr.Logger) {
+
+	lg.V(logs.DebugLevel).Info("Error Event for Secret: " + secret.Name + " " + err.Error())
+
+	tryDeleteEvent(ctx, secret.Name, secret.Namespace, r, lg)
+
+	secretErrEvent := &corev1.Event{}
+	secretErrEvent.Name = secret.Name
+	secretErrEvent.Message = err.Error()
+	secretErrEvent.Namespace = secret.Namespace
+	secretErrEvent.Reason = "Can not upload access token"
+	secretErrEvent.InvolvedObject = corev1.ObjectReference{Namespace: secret.Namespace, Name: secret.Name, Kind: secret.Kind, APIVersion: secret.APIVersion}
+	secretErrEvent.Type = "Error"
+	secretErrEvent.LastTimestamp = metav1.NewTime(time.Now())
+
+	err1 := r.Create(ctx, secretErrEvent)
```

Comment:
better var name

Reply:
:) well agree in general but why err is fine and err1 is not?

Reply:
OK, I change the logic to not to have err1 variable at all for simplicity